bin/ecompress

```diff
@@ -0,0 +1,147 @@
+		if [[ ${skip_dirs[@]} ]]; then
+			find "${skip_dirs[@]}" -name '*.ecompress' -delete || die
+		fi
+
+		exit 0
+		;;
+	--queue)
+		shift
+
+		find_args=()
+		for path; do
+			if [[ -e ${ED%/}/${path#/} ]]; then
+				find_args+=( "${ED%/}/${path#/}" )
+			fi
+		done
+
+		if [[ ${find_args[@]} ]]; then
+			find_args+=( -type f )
+			[[ -n ${PORTAGE_DOCOMPRESS_SIZE_LIMIT} ]] &&
+				find_args+=( -size "+${PORTAGE_DOCOMPRESS_SIZE_LIMIT}c" )
+
+			while IFS= read -d '' -r path; do
+				>> "${path}.ecompress" || die
+			done < <(find "${find_args[@]}" -print0 || die)
+		fi
+
+		exit 0
+		;;
+	--dequeue)
+		[[ -n ${2} ]] && die "${0##*/}: --dequeue takes no additional arguments"
+		break
+		;;
+	*)
+		die "${0##*/}: unknown arguments '$*'"
+		exit 1
+		;;
+	esac
+	shift
+done
+
+# setup compression stuff
+PORTAGE_COMPRESS=${PORTAGE_COMPRESS-bzip2}
+[[ -z ${PORTAGE_COMPRESS} ]] && exit 0
+
+if [[ ${PORTAGE_COMPRESS_FLAGS+set} != "set" ]] ; then
+	case ${PORTAGE_COMPRESS} in
+		bzip2|gzip)  PORTAGE_COMPRESS_FLAGS="-9";;
+	esac
+fi
+
+guess_suffix() {
+	set -e
+	tmpdir="${T}"/.ecompress$$.${RANDOM}
+	mkdir "${tmpdir}"
+	cd "${tmpdir}"
+	# we have to fill the file enough so that there is something
+	# to compress as some programs will refuse to do compression
+	# if it cannot actually compress the file
+	echo {0..1000} > compressme
+	${PORTAGE_COMPRESS} ${PORTAGE_COMPRESS_FLAGS} compressme > /dev/null
+	# If PORTAGE_COMPRESS_FLAGS contains -k then we need to avoid
+	# having our glob match the uncompressed file here.
+	suffix=$(echo compressme.*)
+	[[ -z $suffix || "$suffix" == "compressme.*" ]] && \
+		suffix=$(echo compressme*)
+	suffix=${suffix#compressme}
+	cd /
+	rm -rf "${tmpdir}"
+	echo "${suffix}"
+}
+
+# figure out the new suffix
+export PORTAGE_COMPRESS_SUFFIX=$(guess_suffix) || die
+
+fix_symlinks() {
+	# Repeat until nothing changes, in order to handle multiple
+	# levels of indirection (see bug #470916).
+	local -i indirection=0
+	while true ; do
+	local something_changed=
+	while read -r -d $'\0' brokenlink ; do
+		[[ -e ${brokenlink} ]] && continue
+		olddest=$(readlink "${brokenlink}")
+		newdest=${olddest}${PORTAGE_COMPRESS_SUFFIX}
+		if [[ "${newdest}" == /* ]] ; then
+			[[ -f "${D%/}${newdest}" ]] || continue
+		else
+			[[ -f "${dir}/${brokenlink%/*}/${newdest}" ]] || continue
+		fi
+		something_changed=${brokenlink}
+		rm -f "${brokenlink}" &&
+		ln -snf "${newdest}" "${brokenlink}${PORTAGE_COMPRESS_SUFFIX}"
+		((ret|=$?))
+	done < <(find "${ED}" -type l -print0 || die)
+
+	[[ -n ${something_changed} ]] || break
+	(( indirection++ ))
+	if (( indirection >= 100 )) ; then
+		# Protect against possibility of a bug triggering an endless loop.
+		eerror "ecompress: too many levels of indirection for" \
+			"'${actual_dir#${ED%/}}/${something_changed#./}'"
+		break
+	fi
+	done
+
+	return ${ret}
+}
+
+export PORTAGE_COMPRESS PORTAGE_COMPRESS_FLAGS
+find "${ED}" -name '*.ecompress' -delete \
+	-exec "${PORTAGE_BIN_PATH}"/ecompress-file {} +
```

Comment:
There will always be some odd packages where parallel compression will provide a significant benefit, so it would be nice to use xargs --max-procs as in 50283f1abb77f0785ab86d41ad70d76df4e399be.

Reply:
Yeah, I'd like to do that as a followup patches. I'm generally thinking we could move the logic outta pngfix and e.g. export `${_XARGS_PARALLEL}` with ready args for all potential users.